boa_engine/src/builtins/promise/mod.rs

```diff
@@ -0,0 +1,731 @@
+
+                        // c. Set promiseCapability.[[Resolve]] to resolve.
+                        promise_capability.resolve = resolve.clone();
+
+                        // d. Set promiseCapability.[[Reject]] to reject.
+                        promise_capability.reject = reject.clone();
+
+                        // e. Return undefined.
+                        Ok(JsValue::Undefined)
+                    },
+                    PromiseCapabilityCaptures {
+                        promise_capability: promise_capability.clone(),
+                    },
+                )
+                .name("")
+                .length(2)
+                .build()
+                .into();
+
+                // 6. Let promise be ? Construct(C, « executor »).
+                let promise = c.construct(&[executor], &c.clone().into(), context)?;
+
+                let promise_capability: &mut Self =
+                    &mut promise_capability.try_borrow_mut().expect("msg");
+
+                let resolve = promise_capability.resolve.clone();
+                let reject = promise_capability.reject.clone();
+
+                // 7. If IsCallable(promiseCapability.[[Resolve]]) is false, throw a TypeError exception.
+                if !resolve.is_callable() {
+                    return context
+                        .throw_type_error("promiseCapability.[[Resolve]] is not callable");
+                }
+
+                // 8. If IsCallable(promiseCapability.[[Reject]]) is false, throw a TypeError exception.
+                if !reject.is_callable() {
+                    return context
+                        .throw_type_error("promiseCapability.[[Reject]] is not callable");
+                }
+
+                // 9. Set promiseCapability.[[Promise]] to promise.
+                promise_capability.reject = promise;
+
+                // 10. Return promiseCapability.
+                Ok(promise_capability.clone())
+            }
+        }
+    }
+}
+
+impl BuiltIn for Promise {
+    const NAME: &'static str = "Promise";
+
+    const ATTRIBUTE: Attribute = Attribute::WRITABLE
+        .union(Attribute::NON_ENUMERABLE)
+        .union(Attribute::CONFIGURABLE);
+
+    fn init(context: &mut Context) -> Option<JsValue> {
+        let _timer = Profiler::global().start_event(Self::NAME, "init");
+
+        ConstructorBuilder::with_standard_constructor(
+            context,
+            Self::constructor,
+            context.intrinsics().constructors().promise().clone(),
+        )
+        .name(Self::NAME)
+        .length(Self::LENGTH)
+        .method(Self::then, "then", 1)
+        .build()
+        .conv::<JsValue>()
+        .pipe(Some)
+    }
+}
+
+struct ResolvedRecord {
+    value: bool,
+}
+
+struct ResolvingFunctionsRecord {
+    resolve: JsValue,
+    reject: JsValue,
+}
+
+#[derive(Debug, Trace, Finalize)]
+struct RejectResolveCaptures {
+    promise: JsObject,
+    already_resolved: JsObject,
+}
```

Comment:
You can move this inside `create_resolving_functions`. This should limit its scope, since it's only used internally, and it would make it renameable to only `Captures`.

Reply:
This is also being used in job.rs, so I left it as-is in #2107. Feel free to review it there.